### PR TITLE
Typescript port

### LINKED ISSUE #8: Add TypeScript type declarations?
Hi. Thank you for your library.

Would you be interested in adding [type declarations for TypeScript](https://www.typescriptlang.org/docs/handbook/2/type-declarations.html) to this project? This would make the usage of the library type-safe in TypeScript projects.

If so I could create a PR for this. If not I will create a PR for it in the [DefinitelyTyped repo](https://github.com/DefinitelyTyped/DefinitelyTyped) and users will have to install the type declarations separately.

Adding the type declarations to this repo is the preferred solution.

### PR DESCRIPTION
**Pull request by krisztianb**
Resolves #8 

Changes:

* Code ported to **TypeScript** (including some minor changes)
* Added comments for the methods and properties
* New scripts for build and test
* Updated **jasmine-node**
* Added `file` entry to `package.json` in order not to publish the tests to NPM

Remarks:

* No breaking changes for users. Existing code using the library should still work the same.
* Tests are still passing
* Tested in browser successfully
* Tested in a node.js application successfully
* Code is now in `src/utfstring.ts` which can be compiled using `npm run clean-build`

Let me know what you think.

### CODE REVIEW

File: src/utfstring.ts [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+
+        sources.push(surrogatePairs.source);
+        sources = sources.concat(extraSources);
+
+        return new RegExp(sources.join("|"), modifiers);
+    }
+}
+
+const visual = new UtfStringClass([regionalIndicatorPairs], undefined);
+const UtfString = new UtfStringClass([], visual);
```
Comment (sant123) [inline src/utfstring.ts:447]: I think there is currently missing two exports from here https://github.com/camertron/utfstring/blob/master/utfstring.js#L332

Also do you think the export could be done this way instead?

```ts
export const ...
```

Comment (krisztianb) [reply src/utfstring.ts:447]: The export is done just like it was in the former JavaScript version. The `export` object is overwritten a few lines below those `const` assignments with the `UtfString` constant. So import works the same way and existing code using this library doesn't have to be modified at all.

File: README.md [hunk 2]
```diff
@@ -25,7 +25,7 @@ Encoding is the process of converting Unicode codepoints into binary data that c
 
 #### Javascript's Solution
 
-To mitigate this problem, Javascript (as well as other languages and platforms that use UTF-16 encoding) makes use of what are called "surrogate pairs." Surrogate pairs are two encoded characters that represent a single logical character. Together they are 4 bytes wide and can represent every Unicode character (2<sup>32</sup> = 4,294,967,296).
+To mitigate this problem, Javascript (as well as other languages and platforms that use UTF-16 encoding) makes use of what are called "surrogate pairs". Surrogate pairs are two encoded characters that represent a single logical character. Together they are 4 bytes wide and can represent every Unicode character (2<sup>32</sup> = 4,294,967,296).
```
Comment (camertron) [inline README.md]: It's actually correct for the period to be inside the double quotes. Weird but true!

**Review by camertron — CHANGES_REQUESTED** [review]
Just one minor issue, but this looks really good! Thanks! 🎉

File: README.md (hunk 2 above)
Comment (krisztianb) [reply README.md]: Strange. I can change that back. No problem.

**Review by camertron — APPROVED** [review]
Awesome, thanks!!

**Comment by camertron** [comment]
Realized I never published this! Should be done now, v2.1.0 :)

**Pull request merged**